src/worker/mod.rs

```diff
@@ -369,51 +412,86 @@ impl<
                 .await?;
         }
 
-        let exit = self.listen_for_heartbeats(&statuses).await;
+        // Either we have already taken this future and are polling it elsewhere (see Worker::run_to_completion),
+        // or they've never provided it to us (reminder: they do so via WorkerBuilder::with_graceful_shutdown)
+        let cancel_signal = self.shutdown_signal.take();
+
+        let report = tokio::select! {
+            // A signal from the user space received.
+            _ = async { let signal = cancel_signal.unwrap(); signal.await }, if cancel_signal.is_some() => {
+                let nrunning = tokio::select! {
+                     _ = tokio_sleep(self.shutdown_timeout.unwrap()), if self.shutdown_timeout.is_some() => {
+                        0
+                    },
+                    nrunning = self.force_fail_all_workers("termination signal received from user space") => {
+                        nrunning
+                    }
+                };
+                self.terminated = true;
+                Ok((RunCeaseReason::CancelSignal, nrunning))
+            },
+            // A signal from the Faktory server received or an error occurred.
+            // Even though `Worker::listen_for_hearbeats` is not cancellation safe, we are ok using it here,
+            // since we are not `select!`ing in a loop and we are eventually _either_ marking the worker as
+            // terminated, _or_ re-creating a connection, i.e. we never end up re-using the "broken" connection.
+            exit = self.listen_for_heartbeats(&statuses) => {
+                // there are a couple of cases here:
+                //  - we got TERMINATE, so we should just return, even if a worker is still running
+                //  - we got TERMINATE and all workers has exited
```

Comment:
```suggestion
                //  - we got TERMINATE and all workers have exited
```

Reply:
Hmm I was wondering how this got into the PR, and I remember now that it used to be [`has`](https://github.com/jonhoo/faktory-rs/blob/8c06cc7d0d84d337f2c29da3fab9708cffea143f/src/consumer/mod.rs#L505) and then we corrected it in the make-everything-async merge to be `have`, but I it somehow remained in the initial edition in this branch and so it looks like I am suggesting this, which technically I am :smile: 

I wonder though how you noticed this one among all this whitespace noise :wink: